Arnold5 Experimental For Gaffer 0.35

danieldresser-ie opened the pull request:
This is not quite ready yet, but it's probably about time to start reviewing it.

The one thing I could urgently use review of is 206bc65, since it could be compatibility breaking if we decide to approach this differently.  I know you've historically been hesitant to introduce new plug types, but have a special type for ClosurePlug feels like the clearest way to communicate that we need to connect them to each other, but can't do anything else with them.

If you think I should do this differently, it would be great to hear right away before artists start setting up test data.

Other than that, the main two things that I know are broken currently:
* efb5f43 : This test isn't passing even apart from the crash, but getting an upstream fix for the crash should make this easier to figure out
* 573e939 : I currently skip parameters of type NODE - maybe this OK for the moment?  We need to figure out how hooking up shaders to a skydome_light would work.  Maybe it would be some sort of special case use of ShaderAssignment, and then it would be OK that a generic NODE parameter is ignored.

johnhaddon commented on src/GafferArnold/ArnoldOptions.cpp:
```diff
@@ -100,7 +100,7 @@ ArnoldOptions::ArnoldOptions( const std::string &name )
 	options->addOptionalMember( "ai:texture_searchpath", new IECore::StringData( "" ), "textureSearchPath", Gaffer::Plug::Default, false );
 	options->addOptionalMember( "ai:procedural_searchpath", new IECore::StringData( "" ), "proceduralSearchPath", Gaffer::Plug::Default, false );
-	options->addOptionalMember( "ai:shader_searchpath", new IECore::StringData( "" ), "shaderSearchPath", Gaffer::Plug::Default, false );
+	options->addOptionalMember( "ai:plugin_searchpath", new IECore::StringData( "" ), "pluginSearchPath", Gaffer::Plug::Default, false );
```
Should we add a compatibility config for this?

danieldresser-ie replied:
From an IE point of view, we're not really expecting old scenes to continue working ( and artists aren't overriding this path anyway ).  Do you think there is someone else out there who benefit enough to be worth worrying about?

johnhaddon replied:
I dunno to be honest - if anyone out there would like this, let us know! In the absence of input, perhaps we do nothing for now, knowing we can always implement one after the fact if we find a need?

johnhaddon commented on arnold/plugins/gaffer.mtd:
```diff
@@ -340,1137 +203,602 @@
 
 
-[node userDataVector]
+[node standard_surface]
```
What does the NodeEditor UI for standard_surface look like now? Is it reasonably arranged into sections like the AlSurface was?

danieldresser-ie replied:
Nope, it's a huge mess.  To get it looking nice, I took an mtd for it that ships with htoa, and modified a bunch of things to make it work better with Gaffer, and rewrote some tooltips to make more sense - currently I've only released this internally with IERendering.  Do you think I should be trying to ship this with Gaffer?  I guess I'd have to check with the SolidAngle folks that they're OK with me modifying and redistributing a file that comes with htoa.  ( Also, the syntax of mtd files for Houdini has changed, so if we want to properly be compatible with current mtd files, I guess we should update our mtd parser.  Now that there are only a handful of Arnold shaders we actually want to use, manually rewriting mtd files is quicker than updating the parser, but maybe you feel like this is something we should support in public Gaffer )

johnhaddon replied:
Whatever we do, Gaffer needs to be useable out of the box with nothing more than Arnold core. I think there's a good case to be made that Solid Angle should ship a useable .mtd file with Arnold core, since that's where the shaders are now. I guess they'd be reluctant to include things like the crazy Houdini grouping syntax though unless this has improved. Maybe you could broach the subject with them? Perhaps a compromise is that they ship sensible documentation strings and presets with Arnold core, leaving the crazy weird stuff in HtoA, and then we fill in with some section metadata of our own that maps more directly to Gaffer layouts?

johnhaddon commented on src/GafferArnold/ArnoldOptions.cpp:
```diff
@@ -58,24 +58,23 @@ ArnoldOptions::ArnoldOptions( const std::string &name )
 	options->addOptionalMember( "ai:GI_diffuse_samples", new IECore::IntData( 2 ), "giDiffuseSamples", Gaffer::Plug::Default, false );
-	options->addOptionalMember( "ai:GI_glossy_samples", new IECore::IntData( 2 ), "giGlossySamples", Gaffer::Plug::Default, false );
-	options->addOptionalMember( "ai:GI_refraction_samples", new IECore::IntData( 2 ), "giRefractionSamples", Gaffer::Plug::Default, false );
+	options->addOptionalMember( "ai:GI_specular_samples", new IECore::IntData( 2 ), "giSpecularSamples", Gaffer::Plug::Default, false );
```
Worth adding a compatibility config?

johnhaddon commented on python/GafferArnoldTest/IECoreArnoldPreviewTest/RendererTest.py:
```diff
@@ -1427,15 +1427,14 @@ def testStepSizeAttribute( self ) :
-				"test.so",
+				# There aren't any procedurals we rely on being installed, but we can create any regular
+				# node the same way we create a procedural
+				"cone", 
```
Can we just use "procedural" here?

On a related note, I see that Arnold 5 supports setting "step_size" on polymesh now, so the "convert meshes with a step size to boxes" thing that we're testing here could be removed now. Would be nice to add that before merging.

danieldresser-ie replied:
> Can we just use "procedural" here?

That seems misleading?  If someone actually writes an external procedural, it won't create a node type named "procedural", it will create a node type named "MyCoolThing" or whatever.  The only reason to use "procedural" in an ExternalProcedural is for ass archives, not procedurals.  I guess I could test the ass archive case, but then I should also change the name from "regularProcedural" to "assArchive".

johnhaddon replied:
> That seems misleading?

I don't think it's as misleading as creating a cone!

> The only reason to use "procedural" in an ExternalProcedural is for ass archives, not procedurals. I guess I could test the ass archive case

Sounds good.

> but then I should also change the name from "regularProcedural" to "assArchive".

OK.

danieldresser-ie replied:
In the end, I double checked, and we can get rid of the mesh->box step size thing completely, so I just removed the troublesome line about testing a default procedural, since I guess we're no longer concerned about screwing with procedurals by turning them into meshes.

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -653,9 +653,9 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
 			{
 				const IECore::ExternalProcedural *procedural = static_cast<const IECore::ExternalProcedural *>( object );
-				if( const IECore::StringData *nodeType = procedural->parameters()->member<const IECore::StringData>( "ai:nodeType" ) )
+				if( procedural->getFileName() == "volume" )
```
Don't suppose you could make a super-quick PR for Cortex master, adding a todo to rename the filename property to account for the usage in Arnold? I'll do the actual work when I'm doing some other refactoring later on...

danieldresser-ie replied:
Will do this when I make this change in Cortex master.

johnhaddon commented on src/GafferArnold/ParameterHandler.cpp:
```diff
@@ -430,6 +430,10 @@ Gaffer::Plug *ParameterHandler::setupPlug( const AtNodeEntry *node, const AtPara
 
+		case AI_TYPE_NODE :
+			// TODO - is it safe to ignore this?
+			// Probably not, we should figure out how to hook up shader to skydome_light
```
When I say, "before we merge", I suppose what I really mean is "before this makes it to master". If it's useful to get something less complete onto the experimental branch then that's fine, but we'll need to get the remaining stuff taken care of before merging to master.

danieldresser-ie commented on src/GafferArnold/IECoreArnoldPreview/ShaderAlgo.cpp:
```diff
@@ -123,7 +123,7 @@ std::vector<AtNode *> convert( const IECore::ObjectVector *shaderNetwork, const
 			if( boost::starts_with( shader->getType(), "osl:" ) )
 			{
-				nodeType = "osl_shader";
+				nodeType = "osl";
```
I did note that - it seems to have been there from the start of the prototype osl_shader node as well.

It seems like given OSL's support in the backend for on the fly compiled code, it would be reasonable to expect the renderer backend to expose something like this.

danieldresser-ie commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -1371,7 +1371,8 @@ class ArnoldObject : public IECoreScenePreview::Renderer::ObjectInterface
 		void applyTransform( AtNode *node, const Imath::M44f &transform )
 		{
-			AiNodeSetMatrix( node, "matrix", const_cast<float (*)[4]>( transform.x ) );
+			AiNodeSetMatrix( node, "matrix", *reinterpret_cast<AtMatrix*>(
```
Roger.

danieldresser-ie commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -527,6 +527,17 @@ IECore::InternedString g_curvesModeAttributeName( "ai:curves:mode" );
+const AtString g_boxArnoldString("box");
+const AtString g_volumeArnoldString("volume");
+const AtString g_sphereArnoldString("sphere");
+const AtString g_subdiv_typeArnoldString("subdiv_type");
```
OK - I felt like these might be an exception because the meaning of this name is literally 'a constant which stores the letters "subdiv_type" as an Arnold string', but I can switch them if it bugs you.

danieldresser-ie commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -1378,23 +1379,18 @@ class ArnoldObject : public IECoreScenePreview::Renderer::ObjectInterface
-				AiNodeSetArray( node, "time_samples", timesArray );
-			}
+
+			NodeAlgo::ensureUniformTimeSamples( times );
```
Hmm, if you're generating scenes using Gaffer, it should be impossible to generate that case, right?  In fact, it should be impossible to generate any case that fails ensureUniformSampleTimes either, using Gaffer.

But I guess since we still theoretically support standalone Cortex I should see if there's any way I could flag this without too much difficulty.

johnhaddon replied:
Yeah, Gaffer should be fine, but in theory anyone could use the renderer backends - one outcome of the Encapsulate work is that we will be able to support Python procedurals in Cortex 10 for any die hards that want them (albeit with the new API rather than the old one).

You know what though, I don't see a way of implementing this warning without storing additional state, so unless you can see a way of doing it without overhead, let's not worry about it.

danieldresser-ie commented on src/GafferArnold/ParameterHandler.cpp:
```diff
@@ -437,10 +437,11 @@ Gaffer::Plug *ParameterHandler::setupPlug( const AtNodeEntry *node, const AtPara
 				AiParamGetName( parameter ) %
 				AiParamGetTypeName( AiParamGetType( parameter ) ) %
-				nodeName( plugParent )
+				nodeName( plugParent ) %
```
More helpful than nothing in an error message - if you have ideas about how you'd prefer to see it formatted, I'm open.

danieldresser-ie commented on src/GafferArnold/ArnoldAttributes.cpp:
```diff
@@ -54,10 +54,12 @@ ArnoldAttributes::ArnoldAttributes( const std::string &name )
-	attributes->addOptionalMember( "ai:visibility:refracted", new IECore::BoolData( true ), "refractedVisibility", Gaffer::Plug::Default, false );
-	attributes->addOptionalMember( "ai:visibility:diffuse", new IECore::BoolData( true ), "diffuseVisibility", Gaffer::Plug::Default, false );
-	attributes->addOptionalMember( "ai:visibility:glossy", new IECore::BoolData( true ), "glossyVisibility", Gaffer::Plug::Default, false );
+	attributes->addOptionalMember( "ai:visibility:diffuseReflect", new IECore::BoolData( true ), "diffuseReflectionVisibility", Gaffer::Plug::Default, false );
```
I don't see any particular utility in a compatibility config.  The functionality of the renderer has changed pretty substantially when it comes to classifying rays, and I don't think we should try and create any impression that old quality settings will be meaningful in Arnold 5 renders.

danieldresser-ie commented on include/GafferScene/ClosurePlug.h:
```diff
@@ -0,0 +1,77 @@
+/// loader a shader from OSL or a renderer.  We probably won't be able
+/// to set or get closure plugs, but we need to be able to connect 
+/// them, and they should only connect to other closure plugs.
+class ClosurePlug : public Gaffer::Plug
```
I think I mentioned this in an earlier email, but yeah, this was the part I was most uncertain in.  It isn't any different than the OSL case, and if we decide to use a specific plug type here, then we should definitely use it for OSL as well ( I've now tested, and it is kosher in Arnold to connect Arnold closure -> OSL or OSL closure -> Arnold ).

As you point out, we have managed without it in OSL so far, so we definitely don't need it ( though I'm not sure if anyone has actually used OSL closures in Gaffer much yet ).  The main reason I went this way to start with is that I had no idea which was better and I needed to get something released for testing ASAP, and it seemed easier to do a compatibility config later to remove the custom plug type than to add it if we decided it was desirable.

Talking with Don, he thought that a custom plug type made things a lot clearer, and I agree.  It is a special type of data and closure plugs should only ever be allowed to connect to other closure plugs.  We can achieve this by using a default Plug type and putting a custom acceptsInput on any node that ever has a closure plug on it, but this seems like a complex and failure prone way to achieve the same thing as actually having a ClosurePlug type?

( Oh, and the shaderOutPlug was just a sloppy copy/paste while I was setting up this prototype ).

danieldresser-ie replied:
Oh, and also, looking at the current logic in OSLShader acceptsInput, it currently allows you to connect any raw Plug into a closure Plug?  That seems like it could cause problems if anything in Gaffer uses Gaffer::Plug to mean anything other than a closure plug?

danieldresser-ie commented:
OK, I think I've covered all the little comments.  Haven't added compatibility configs yet, but it doesn't sound like they're too necessary.

While cleaning up Arnold attribute names, I also added a commit to rename ai:polymesh:subdividePolygons to ai:polymesh:subdivide_polygons, since that seems to be the convention you want?

For the closure plug stuff, for the moment I've switched OSL over to using ClosurePlug as well.  If you want everything to just use Plug, this should be an easy switch ( though I still think a step backward ).

And still remaining:
* The shader parameter of type NODE on skydome light is documented here: https://support.solidangle.com/display/A5NodeRef/skydome_light
It's too bad that if their intention was to allow you to connect closures, they didn't make it of type closure ( like other shaders that accept closures ), rather than node, but I guess to them, a light is more a shape, whereas to us it's more like a shader.  Actually, I just realized:  if I load NODE parameters on lights as ClosurePlugs, I think everything might just work - gonna try that first thing tomorrow.

* Still haven't sorted out the trace_set thing.  I might hack around the crash to get the test actually working ( the test occurs after the render finishes, which is annoying but I should be able to work around it enough to confirm that we can use trace sets properly ).

johnhaddon commented on src/GafferOSL/OSLShader.cpp:
```diff
@@ -604,13 +600,13 @@ Plug *loadMatrixArrayParameter( const OSLQuery::Parameter *parameter, const Inte
 {
-	Plug *existingPlug = parent->getChild<Plug>( name );
+	ClosurePlug *existingPlug = parent->getChild<ClosurePlug>( name );
 	if(	existingPlug && existingPlug->typeId() == Plug::staticTypeId() )
```
This test is going to fail now, and we'll end up replacing plugs that were perfectly good. I think now it just needs to be the following :

```
if( existingPlug )
{
    return existingPlug;
}
```

johnhaddon commented on python/GafferArnoldTest/IECoreArnoldPreviewTest/RendererTest.py:
```diff
@@ -943,7 +943,7 @@ def testSubdividePolygonsAttribute( self ) :
 			a = IECore.CompoundObject()
 			if subdividePolygons is not None :
-				a["ai:polymesh:subdividePolygons"] = IECore.BoolData( subdividePolygons )
+				a["ai:polymesh:subdivide_polygons"] = IECore.BoolData( subdividePolygons )
```
There was method in the madness of the existing naming - "subdividePolygons" isn't a native Arnold attribute but one we synthesise ourselves, so it was following the Cortex convention as a small reminder of that fact. I think we discussed this way back when, but maybe you feel this is definitely better?

danieldresser-ie replied:
Well, the thing is that diffuse_reflect isn't a native Arnold name either - we use it to set a bitmask flag named AI_RAY_DIFFUSE_REFLECT, so to me both subdividePolygons/diffuseReflect or subdivide_polygons/diffuse_reflect are equally names we're synthesizing ourselves, and I feel like they ought to be consistent?

johnhaddon replied:
True, although DIFFUSE_REFLECT does at least map directly to an Arnold concept of the same name, whereas subdivide_polygons isn't even really an Arnold concept. Still, I don't feel like I'm on very strong ground here, so if you think this is definitely better that's fine with me...

johnhaddon commented:
> For the closure plug stuff, for the moment I've switched OSL over to using ClosurePlug as well. If you want everything to just use Plug, this should be an easy switch ( though I still think a step backward ).

No, I agree that ClosurePlug is better. Every time I've tried to recycle a plug type but give it slightly different semantics, I've regretted it in the long run. I've made a note of a bug on your OSLShader commit though, and I notice you haven't updated `OSLShader::acceptsInput()` to take account of the change either. Note also that you'll need ClosurePlug to do something akin to [ShaderPlug](https://github.com/GafferHQ/gaffer/blob/master/src/GafferScene/ShaderPlug.cpp#L104) to deal with closures that have been promoted in the past.

Also, I think ClosurePlug should be in GafferOSL rather than GafferScene, since I don't think we can expect there to be other closure systems that will merrily interoperate with each other. For instance, I don't think PRMan supports OSL closures yet, but does have a separate non-compatible concept of its own.

> The shader parameter of type NODE on skydome light is documented here: https://support.solidangle.com/display/A5NodeRef/skydome_light

Ah cool, thanks for the reference. So it sounds like we can still make connections to the color if we want, so we wouldn't be losing anything from the functionality we had before?

> It's too bad that if their intention was to allow you to connect closures, they didn't make it of type closure ( like other shaders that accept closures ), rather than node, but I guess to them, a light is more a shape, whereas to us it's more like a shader.

Yeah, it does seem a bit odd, but I guess its consistent with polymesh.disp_map.

> Actually, I just realized: if I load NODE parameters on lights as ClosurePlugs, I think everything might just work - gonna try that first thing tomorrow.

I suspect that will work, but maybe it'd be making assumptions that wouldn't hold for other parameters in the future? What do you think to supporting it by adding a "gaffer.plugType" metadata entry saying explicitly that the skydome_light.shader parameter should be represented as a ClosurePlug? We've used that strategy elsewhere, and it seems like it might be better than guessing...

danieldresser-ie commented:
Cool - I'm fine with putting ColorPlug in GafferOSL.

 ( As an aside, I'm also not completely convinced that we would need a separate type for PRMan.  Currently, a PRMan utility shader and a Arnold utility shader could both have plugs of type ColorPlug, and we would currently allow you to connect them - it doesn't seem like plug type needs to be responsible for everything, and to me ColorPlug in PRMan vs ColorPlug in Arnold seems to have the same relationship as ClosurePlug in PRMan vs ClosurePlug in Arnold.  But this would be pretty easy to change later, I'm fine with putting ClosurePlug in GafferOSL for the moment. )

> I've made a note of a bug on your OSLShader commit though, and I notice you haven't updated OSLShader::acceptsInput() to take account of the change either.

Roger

> Note also that you'll need ClosurePlug to do something akin to ShaderPlug to deal with closures that have been promoted in the past.

Wait, how much old compatibility are we worried about here?  This isn't an issue at IE, so it's only in the future public Gaffer 0.40.0 that this could be a concern.  Do we have a hard rule that Gaffer doesn't break scene compatibility even on fairly major releases?  Something like the UV orientation flip is going to break possible Gaffer scenes anyway.  The people we're worried about are basically just Appleseed users?  Would it be worth checking if any Appleseed users have existing scenes they are worried about preserving?  I'm fine with doing the work to add compatibility stuff, but I would have hoped that this release cycle would be a time when we could remove stuff like that deprecated support in ShaderPlug.

> What do you think to supporting it by adding a "gaffer.plugType" metadata entry saying explicitly that the skydome_light.shader parameter should be represented as a ClosurePlug?

I'll try this out.

johnhaddon commented:
> As an aside, I'm also not completely convinced that we would need a separate type for PRMan. Currently, a PRMan utility shader and a Arnold utility shader could both have plugs of type ColorPlug, and we would currently allow you to connect them - it doesn't seem like plug type needs to be responsible for everything, and to me ColorPlug in PRMan vs ColorPlug in Arnold seems to have the same relationship as ClosurePlug in PRMan vs ClosurePlug in Arnold. But this would be pretty easy to change later, I'm fine with putting ClosurePlug in GafferOSL for the moment.

That's an interesting argument I hadn't considered. Still, I think putting it in GafferOSL at first is the safest option - it should be trivial to promote it out if it makes sense when we get to adding PRMan support.

>>Note also that you'll need ClosurePlug to do something akin to ShaderPlug to deal with closures that have been promoted in the past.
>Wait, how much old compatibility are we worried about here? This isn't an issue at IE, so it's only in the future public Gaffer 0.40.0 that this could be a concern.

Perhaps it isn't obvious, but all the OSL shaders for use with OSLObject and OSLImage pass closures, so there are for sure closure plugs (as Plugs) in the wild at IE.

> Do we have a hard rule that Gaffer doesn't break scene compatibility even on fairly major releases?

I would say we have a relatively hard rule that we balance with pragmatism and intuition about how likely problems are in the wild. In this case I don't think it's a big deal to add compatibility, and I think there's a decent chance we need it.

> Something like the UV orientation flip is going to break possible Gaffer scenes anyway.

Actually, it only breaks scenes for folks who have been manipulating UVs within Gaffer. For more regular lookdev/lighting scenarios it should be transparent. But the bigger point here would be that to get to where we want to be with UVs we _have_ to break something. Here compatibility would be trivial.

> I would have hoped that this release cycle would be a time when we could remove stuff like that deprecated support in ShaderPlug.

Yeah, that's not unreasonable. Since we're not talking about a lot of work though, I'd be more comfortable if the Arnold 5 upgrade didn't affect orthogonal workflows (OSLObject), and we made a more explicit break separately.

danieldresser-ie commented:
OK, got pulled off onto some other things ( users actually testing this stuff, hurrah ), and didn't get much done today aside from tinkering with ClosurePlug.

I think it's OK now, aside from:
* This cast to int is kinda sketchy, to get around the fact that we're now switching over typeIds from two different enums ( maybe you have a better solution ) ( oh also GafferArnold now depends on GafferOSL, but I guess that's cool ):  
https://github.com/danieldresser-ie/gaffer/blob/3c09b530e2cc1c387dabddf7781ccd4bb36b9649/src/GafferOSL/OSLCode.cpp#L87

* I haven't done the compatibility fallback for ClosurePlug::acceptsInput yet.  Things I realized after talking to you:
 1) we probably wouldn't ever actually hit it at IE, because when using OSLObject, you connect the closure directly to OutObject, and the OutPoint and OutObject nodes stay stuck together, and you probably wouldn't promote just one of them or put a dot in between
 2) if we do have any setups that would be affected by this at IE, now is my best opportunity to make a compatibility breaking change in the next year, so all the more reason why I would really like to not put compatibility in place for 0.3X_experimental_ai5 at least
 3) I really don't like the compatibility mode on ShaderPlug, though it may be necessary at this point.  ShaderPlug::acceptsInput says at the end that it can't return false for reasons of backwards compatibility, but at this point, there are actually two important reasons it needs the fallback that are not related to backwards compatibility:  accepting something like a ColorPlug input promoted on a Box ( which could be subsequently connected to a shader out plug ), and OSL surface shaders currently use an output of type Plug ( this probably should be ShaderPlug somehow? )

johnhaddon commented:
> OK, got pulled off onto some other things ( users actually testing this stuff, hurrah ), and didn't get much done today aside from tinkering with ClosurePlug.

No worries - I'll wait until you're ready before going over the code again then.

> This cast to int is kinda sketchy, to get around the fact that we're now switching over typeIds from two different enums ( maybe you have a better solution )

Maybe we could ditch the cast in the `switch()` bit, and instead use `case ClosurePlug::staticTypeId() :` lower down. That isolates the slight oddity to the exact place it's needed, and the cast will be hidden inside the `staticTypeId()` call.

As an aside, I'd love to ditch TypeIds and just use the type names instead. The main benefit of ids was fast comparisons, but since then we've introduced InternedString, and that has fast comparisons anyway (it just compares pointers). So if we used InternedStrings for typeNames, we might be able to live without typeIds. I don't dare embark on such low level changes just now, but I do think it would be a pretty substantial improvement since it would remove the need of managing central lists of ids.

> ( oh also GafferArnold now depends on GafferOSL, but I guess that's cool ):

Yeah, that's cool. GafferArnoldUI depends on GafferOSL already, for the gobo preview.

> I haven't done the compatibility fallback for ClosurePlug::acceptsInput yet...we probably wouldn't ever actually hit it at IE...if we do have any setups that would be affected by this at IE, now is my best opportunity to make a compatibility breaking change in the next year, so all the more reason why I would really like to not put compatibility in place for 0.3X_experimental_ai5 at least

OK, on your head be it and all that...

> I really don't like the compatibility mode on ShaderPlug, though it may be necessary at this point. ShaderPlug::acceptsInput says at the end that it can't return false for reasons of backwards compatibility, but at this point, there are actually two important reasons it needs the fallback that are not related to backwards compatibility: accepting something like a ColorPlug input promoted on a Box ( which could be subsequently connected to a shader out plug ),

This is a fair point, but I don't think it's quite as bad as you're making out. `ShaderPlug::acceptsInput()` will be called again when the input to the promoted ColorPlug is made, so the promoted plug can't be used as a pathway to making an invalid connection into the ShaderPlug.

> and OSL surface shaders currently use an output of type Plug ( this probably should be ShaderPlug somehow? )

I don't see any real benefit in that. ShaderPlug is intended to encapsulate the requirements of nodes that _receive_ shaders as inputs, rather than nodes that _output_ them. I don't see any benefit in some shader nodes using a ShaderPlug output while others (by necessity) continue to use ColorPlug or FloatPlug outputs. We'll still always need ShaderPlug to accept non-ShaderPlug inputs provided they are [valid shader outputs](https://github.com/GafferHQ/gaffer/blob/master/src/GafferScene/ShaderPlug.cpp#L93), and it's clearer to stick to using ShaderPlug for node inputs, not not shader outputs.

danieldresser-ie commented:
> This is a fair point, but I don't think it's quite as bad as you're making out. ShaderPlug::acceptsInput() will be called again when the input to the promoted ColorPlug is made, so the promoted plug can't be used as a pathway to making an invalid connection into the ShaderPlug.

Yeah, it's maybe a reasonable approach - I think at this point my problem is maybe more just the documentation, which currently implies this is there primarily for backwards compatibility, which isn't true.  If we were going to do something for ClosurePlug that was actually for backwards compatibility, we would want to be more restrictive - ClosurePlugs should not accept inputs from ColorPlugs on boxes.  But maybe for the moment you'll let me get away with forcing users to clean up the few networks that might be using Plugs for ClosurePlugs currently.

> I don't see any real benefit in that...

Yeah, sorry, I realized after walking home that this second point was total nonsense.

johnhaddon commented:
> If we were going to do something for ClosurePlug that was actually for backwards compatibility, we would want to be more restrictive - ClosurePlugs should not accept inputs from ColorPlugs on boxes.

Correct - they should only accept inputs from Plugs (not derived classes, just Plugs).

> But maybe for the moment you'll let me get away with forcing users to clean up the few networks that might be using Plugs for ClosurePlugs currently.

OK, let's see how that works out - we can always add it back in later if needs be. Could you make sure to add a "Breaking Change" comment to your OSLShader commit message though, so we can document it in the release notes?

danieldresser-ie commented:
> Maybe we could ditch the cast in the switch() bit, and instead use case ClosurePlug::staticTypeId() : lower down. That isolates the slight oddity to the exact place it's needed, and the cast will be hidden inside the staticTypeId() call.

I kind of thought that the case had to be a constant, but assumed you knew something I didn't here ... but the compiler just gives me ```error: a function call cannot appear in a constant-expression``` if I try this.

johnhaddon commented:
> I kind of thought that the case had to be a constant, but assumed you knew something I didn't here ... but the compiler just gives me error: a function call cannot appear in a constant-expression if I try this.

Ah, yep, oversight on my part. I think we should be able to declare `staticTypeId()` and `staticTypeName()` as `constexpr` now we've moved to C++11, but let's not get this PR tangled up with that. I still think it's worth putting the cast next to the problematic case rather than at the top though, so how about `static_cast<IECore::TypeId>( ClosurePlugTypeId )`?

danieldresser-ie commented:
That still produces:
```error: case value '110980' not in enumerated type 'const Gaffer::TypeId' [-Werror=switch]```
I guess switch statements on enums perform a special check that all the cases are valid enum values.  Don't really see any way around this other than the cast to int.

danieldresser commented:
Hey John, I should mention what I was in the middle of investigating on Friday, because you might be able to tell me the right approach: Arnold subsurface tracesets are driven by a special geometry attribute "sss_traceset" that has to be declared as user data.  This was on my list of things to make sure we support properly in Arnold 5 - it appears we didn't ever support it properly in Arnold 4, but I might as well still do it.  It seems pretty simple, but I'm not sure which approach would be considered cleanest.

The two possibilities for this:
 * hard-coded support for ai:sss_traceset.  The annoying bit is that we would need special case code for deleting the user attribute if sss_traceset is removed in IPR.
 * support "ai:declare:" for attributes the same way as we do for options.  Simplest might be if attributes declared this way were added to IECoreScenePreview::Renderer::ArnoldAttributes::m_user, but with the prefix removed.  Potentially then this loop that removes stale user attributes: https://github.com/GafferHQ/gaffer/blob/master/src/GafferArnold/IECoreArnoldPreview/Renderer.cpp#L755 could loop over all attributes, not just ones prefixed with "user:"?

johnhaddon commented:
> I guess switch statements on enums perform a special check that all the cases are valid enum values. Don't really see any way around this other than the cast to int.

Yep, looks like it. Thanks for exploring the other options though.

> Arnold subsurface tracesets are driven by a special geometry attribute "sss_traceset" that has to be declared as user data.

Just so I understand correctly, does setting this attribute make Arnold do the trace set thing automatically? If so, it isn't really user data in the traditional sense, is it? User data isn't expected to influence the renderer, just to be available for query from shaders...

Presumably it doesn't have any unwanted interaction with instancing? i.e. It's OK to have two ginstances in different sets, both referencing the same polymesh?

> hard-coded support for ai:sss_traceset. The annoying bit is that we would need special case code for deleting the user attribute if sss_traceset is removed in IPR.

The benefits of this seem to be that it makes it clearer that this is a genuine Arnold feature we're mapping to, not just some arbitrary user data. It makes it feel more like a valid citizen of ArnoldAttributes, which is where I presume we'd want to present it. The annoying bit seems a bit annoying, but we're still talking a trivial amount of code I think?

> support "ai:declare:" for attributes the same way as we do for options.

Seems like this might have other uses - do you have any in mind?

I think my preference is probably the "ai:sss_traceset" approach, just because it presents the feature to the user in the same way as Arnold's other features, and it doesn't seem like too much effort to do it this way.

johnhaddon commented:
> Do you want me to run tests to see if anything like this works in OSL in Arnold?

Nah, if you're happy enough with making an "ai:sss_traceset" attribute then that'll do for me.

danieldresser-ie commented:
OK, I think this might be about good to go.

Lots and lots of little commits, but most of them are actually pretty independent.  Let me know if you want me to squash things together.

I'm trying to think of things we haven't talked about yet that you might want to check specifically ...
* I store transform_type as an int.  Storing Arnold enums as a string is more self-documenting, but potentially doing string conversions on every ginstance felt wasteful.
* In ArnoldShaderUI, if no label is given, I now translate Arnold's snake_case parameters into nice labels automatically ( saved having to manually add labels for everything in standard_surface ).
* I've add enough organization metadata to standard_surface to make it not too horrific to use.

I think I'll do another test release tomorrow.  If this is looking OK, it would be good to get #2292 merged to master, because I'd like to backport it to this branch before I release ( if there's more iteration needed on this stuff, I'll just do a test release from my working branch ).

johnhaddon commented on python/GafferArnoldUI/ArnoldAttributesUI.py:
```diff
@@ -239,6 +251,27 @@ def __volumeSummary( plug ) :
+
+			"description",
+			"""
+			Choose how transform motion is interpolated.  Linear produces classic Renderman-style linear vertex motion, rotate_about_origin produces curves arcs centerd on the object's origin, and rotate_about_center, the default, produces curved arcs centred on the object's bounding box middle.
```
- Let's drop the reference to RenderMan - there was nothing in the standard to require linear motion, some renderers did curved motion, and the standard isn't even standard any more.
- I presume you mean "curved arcs" rather than "curves arcs", although maybe one of those words is redundant anyway?
- "centerd" -> "centred"
- But of line wrapping?

johnhaddon commented on python/GafferArnoldUI/ArnoldAttributesUI.py:
```diff
@@ -239,6 +251,27 @@ def __volumeSummary( plug ) :
+		],
+		"attributes.transformType.value" : [
+
+			"presetNames", lambda plug : IECore.StringVectorData( __transformTypeEnumNames ),
```
Quick comment for when we switch to string values. Elsewhere we've used Gaffer-style CamelCase for the enum names and then "arnold_style" for the values. Let's do the same here.

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -1139,6 +1143,7 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
 		IECore::ConstObjectVectorPtr m_lightShader;
 		std::vector<ArnoldShaderPtr> m_lightFilterShaders;
 		IECore::ConstInternedStringVectorDataPtr m_traceSets;
+		int m_transformType;
```
I doubt there's any real performance benefit to storing this as an int rather than an InternedString - we have much bigger fish to fry. But if you really want to do it, we should use the AiEnum API to convert from the string attribute to the internal int, so we'll naturally be in sync with any enum value changes in the future.

danieldresser-ie replied:
I'm just nervous because Arnold have explicitly deprecated all functions in their API that take char* instead of their own interned string called AtString, because they are worried about the performance impact of implicit conversions to AtString.  I might try out the AiEnum API approach.

johnhaddon commented on python/GafferArnoldUI/ArnoldShaderUI.py:
```diff
@@ -178,9 +178,15 @@ def __translateNodeMetadata( nodeEntry ) :
+			# Label from Arnold naming convention
+			# Arnold uses snake_case rather than camelCase for naming, so translate this into
+			# nice looking names
+			__metadata[paramPath]["label"] = " ".join( [ i.capitalize() for i in paramName.split( "_" ) ] )
```
What happens to names without "_" in them? It seems like they'll remain uncapitalised?

danieldresser-ie replied:
For names with a single token, they match Gaffer's camelCase naming convention, and Gaffer's autogenerated labels will be correct anyway.  Do you want me to generate labels for them as well for consistency, even though they won't change anything?

johnhaddon replied:
Ah, of course. No, that's fine - I'd just missed the obvious.

johnhaddon commented on python/GafferArnoldUI/ArnoldAttributesUI.py:
```diff
@@ -70,6 +70,9 @@ def __shadingSummary( plug ) :
 			info.append( IECore.CamelCase.toSpaced( childName ) + ( " On" if plug[childName]["value"].getValue() else " Off" ) )
 
+	if plug["sssSetName"]["enabled"].getValue() :
+		info.append( "Sss Set Name " + plug["sssSetName"]["value"].getValue() )
```
"SSS Set Name", capitalising the acronym bit? Might need to do the same with an explicit label metadata too?

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -584,6 +585,7 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
 
 			m_linkedLights = attribute<IECore::StringVectorData>( g_linkedLights, attributes );
 
+			m_sssSetName = NULL;
```
This is doubly redundant. intrusive_ptr initialises to null by default anyway, and it's also explicitly initialised to null in the initialiser list above.

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -593,7 +595,11 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
 					}
 				}
-				if( boost::starts_with( it->first.string(), g_lightFilterPrefix.string() ) )
+				else if( it->first == g_sssSetNameName )
```
Let's move this out of the loop and do a single lookup with `attribute<IECore::InternedStringVectorData>( g_sssSetNameName, attributes )` above, as we have for all the others. This has the advantage of being less verbose and reporting any invalid data types to the user, and I think it also reduces the number of comparisons...

danieldresser-ie commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -1836,6 +1860,14 @@ class ArnoldRenderer : public IECoreScenePreview::Renderer
 				try
 				{
 					m_outputs[name] = new ArnoldOutput( name, output, nodeDeleter( m_renderType ) );
+					if( m_outputs[name]->lpeValue().size() )
```
Hmm, yeah, I should only be generating the explicit list of lpes into temp data - that totally makes sense, especially now that the ArnoldOutput has a convenient way to grab the lpe, which wasn't the case when I first wrote it.  Should be an easy clean up.

johnhaddon replied:
While you're at it, do we need to return `lpeName()` and `lpeValue()` separately and then do the formatting outside of ArnoldOutput? The previous design had everything encapsulated so that `ArnoldOutput` was the only thing that really knew the details of outputs and their formatting, and then ArnoldRenderer just had to build the final parameter arrays. Maybe we should replace the `ArnoldOutput::string()` and `ArnoldOutput::lpe*()` methods with a single `ArnoldOutput::append( vector<string> &outputs, vector<string> &light_path_expressions )` method? It's a minor thing, but I think it's a worthwhile improvement given that we're rejigging this a bit anyway...

danieldresser-ie commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -343,6 +343,8 @@ class ArnoldOutput : public IECore::RefCounted
 			// supported by Arnold.
 
 			m_data = output->getData();
+			m_lpeName = name.string() + "__LPE__";
```
I need to make sure that LPE names are completely unique.  If you create a display named "diffuse" using an LPE, it would currently reference an LPE named "diffuse__LPE__".  If it tried to reference an LPE named "diffuse", there's a chance Arnold could have a default shader output named "diffuse", and we would get that instead of our LPE.  I'm happy for alternate naming conventions if you have an idea.  I guess usually in Gaffer we would just use a giant hash value to name something that's been generated, but diffuse__LPE__ is more legible in the ass than a MurmurHash.

danieldresser-ie commented on python/GafferArnoldTest/IECoreArnoldPreviewTest/RendererTest.py:
```diff
@@ -536,6 +536,42 @@ def testOutputFilters( self ) :
+			IECore.Display(
+				"beauty.exr",
+				"exr",
+				"lpe:C.*D.* RGB",
```
Oh, right, we currently accept either Arnold naming convention ( "aov RGB" ) , or Gaffer naming convention ( "color aov" ).  Internally, I've ended up mostly just using Arnold naming convention to match the Arnold references, but I guess it makes sense that you would prefer we make things match the Gaffer naming convention instead.

The problem with either "color lpe:C.*" or "lpe C.*" is that neither offers a way to differentiate between the final Arnold data "myLpeName RGB" and "myLpeName RGBA".  I guess it would be somewhat reasonable to always use RGB for LPEs, but there could be legitimate cases where you want to use an LPE for the beauty, and then you would like to get an LPE aov with alpha.

Do we need to start by establishing a general Gaffer convention for aovs which include alpha ( even apart from LPEs )?  Data of "color4 aov"?  Add an Output parameter "includeAlpha" ?

If we had an Output parameter "includeAlpha" to control whether you actually want "lpeName RGB" or "lpeName RGBA", then putting "lpe C.*" in the data sounds fine to me.

johnhaddon replied:
> Internally, I've ended up mostly just using Arnold naming convention to match the Arnold references, but I guess it makes sense that you would prefer we make things match the Gaffer naming convention instead.

That's definitely my preference, yes. In an ideal world, output definitions would be transferrable between renderers, especially since most of them are using OSL now. In practice I don't think we're there yet, because Arnold/PRMan support LPEs but 3Delight supports a separate `outputvariable()` closure, and Appleseed only has a few hardcoded AOVs. But I'd like to aim that way...

> The problem with either "color lpe:C." or "lpe C." is that neither offers a way to differentiate between the final Arnold data "myLpeName RGB" and "myLpeName RGBA". I guess it would be somewhat reasonable to always use RGB for LPEs, but there could be legitimate cases where you want to use an LPE for the beauty, and then you would like to get an LPE aov with alpha.

Ah yep, that makes sense. I do still like "lpe C." though, because it matches the existing "<type> <name>" syntax.

> Do we need to start by establishing a general Gaffer convention for aovs which include alpha ( even apart from LPEs )? Data of "color4 aov"? Add an Output parameter "includeAlpha" ?

3Delight (NSI) has gone with the "includeAlpha" approach, and it seems pretty reasonable to me. From a user perspective, I think a separate checkbox is preferable to mucking about with strings in the data field.

> If we had an Output parameter "includeAlpha" to control whether you actually want "lpeName RGB" or "lpeName RGBA", then putting "lpe C.*" in the data sounds fine to me.

I think we 're in agreement then. Although I suggest we just assume RGB for now, and then add the "includeAlpha" support later, so we can put more time/thought into making it work across all the renderer backends. There's already a tonne of stuff happening in this PR, and I don't think RGB-only LPEs is really a problem for a first release.

danieldresser-ie commented:
OK, aside from the comments I've made, those sound pretty straightforward to address.

danieldresser-ie commented:
OK, all of that seems straightforward other than
* naming LPEs outputName__LPE__ - the name does need to be unique, and I wasn't sure what convention you would prefer.
* for lpe specification naming, I've gone with "lpe C.*".  In order to cover all the possibly needed cases, I've added in a convention of appending ",alpha" to the end of the data if you want alpha.  This is kinda ugly in the case of LPEs, but it matches the old 3delight standard, which I guess is where the rest of the data naming came from.  I now support "color blah,alpha" to output "RGBA blah", and "lpe C.*,alpha" to output "RGBA outputName__LPE__".  If this is what you want, I should squash it in before merging.

johnhaddon commented:
> naming LPEs outputName__LPE__ - the name does need to be unique, and I wasn't sure what convention you would prefer.

For all other namespacing problems like this, we've used "<prefix>:", so I'd suggest something along those lines. We're already using an "ieCoreArnold:" prefix pretty consistently to denote things that are generated by the renderer backend but might otherwise clash with names from elsewhere, so maybe that would make sense here too?

> for lpe specification naming, I've gone with "lpe C.*".

Great.

> In order to cover all the possibly needed cases, I've added in a convention of appending ",alpha" to the end of the data if you want alpha. This is kinda ugly in the case of LPEs, but it matches the old 3delight standard, which I guess is where the rest of the data naming came from. I now support "color blah,alpha" to output "RGBA blah", and "lpe C.*,alpha" to output "RGBA outputName__LPE__". If this is what you want, I should squash it in before merging.

I think that's probably the best we can do in the existing convention, but I think I do prefer the "includeAlpha" parameter idea. Like I said above, I'd actually prefer to leave out the RGBA support for now and be forced to generalise it across all the renderers later. But if you think you _need_ it now, let's stick with what you've got, but on the proviso that it might change...

danieldresser-ie commented:
> I think that's probably the best we can do in the existing convention, but I think I do prefer the "includeAlpha" parameter idea. Like I said above, I'd actually prefer to leave out the RGBA support for now and be forced to generalise it across all the renderers later. But if you think you need it now, let's stick with what you've got, but on the proviso that it might change...

Hmm, any reason why I couldn't just check for an optional "includeAlpha" in the parameters() on the output right now?  I guess in the long run, you would want to make this more visible somehow, but having it in the parameters would at least allow an expert user ( ie. me ) to set it if it became necessary during production for some reason.

johnhaddon commented:
> Hmm, any reason why I couldn't just check for an optional "includeAlpha" in the parameters() on the output right now?

No, I guess not...

danieldresser-ie commented:
OK, I've addressed comments on LPEs ( I went with the includeAlpha parameter, which even if it isn't final, seems like the right direction ), and squashed that.

The only other change was that I added a fix I'd missed to the OSLCodeUI for the switch to ClosurePlug.

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -354,25 +356,45 @@ class ArnoldOutput : public IECore::RefCounted
 		}
 
-		std::string string() const
+		void append( std::vector<std::string> &outputs, std::vector<std::string> &light_path_expressions ) const
```
Camel case please - lightPathExpressions.

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -813,6 +816,16 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
+					// \todo : Arnold quite explicitly discourages constructing AtStrings repeatedly,
+					// but given the need to pass m_transformType around as a string for consistency
+					// reasons, it seems like there's not much else we can do here.
+					// Long run, it seems like a lot of things would get cleaner if we had first
```
This class is destined for Cortex, and has nothing to do with enum plugs at all - it's a whole level lower than that. Why do you say "there's not much else we can do here" - did the AiEnum query idea not work out?

Also, note that the commit message is out of sync with the code here - it's still describing the previous approach. While on the subject of commit messages, it'd be great if you could keep the subject line short enough to avoid line wrapping, and put any extra information in the body - it makes the logs and the review pages much easier to read.

danieldresser-ie replied:
I tried to figure out how the AiEnum approach would work, but it looks like Renderer::ArnoldAtttributes gets repeatedly constructed for each location anyway - as long as the public interface Renderer::attributes gets called for every location, and that takes strings for this property, it seems like we can't really do any better here  ( doing a linear search against a vector of strings probably would be a little bit faster, but would be pretty ugly, and still not particularly optimal ).

The comment on enum plugs comes from talking to Don about it, and him pointing out that aside from performance, a string really isn't a very clean representation of something that can only have 3 values.  Currently, if someone enters this string, and misspells it, we don't flag that until it hits the Arnold ( if I recall correctly, Arnold's error message for this case isn't terrible, but it still seems less than ideal. )  So, anyway, I've kind of given up on anything for the short term, but if this happens often, something cleaner would be nice.

I can scrap the comment if you prefer.  As for mentioning plugs - yes, if we were going to do it, the plug would have to be backed by some sort of Cortex type, and that's what the renderer backend would see, but I don't see much problem mentioning plugs in a comment - is there any point in pretending that anyone cares about Cortex renderer backends in any context other than Gaffer?

danieldresser-ie commented:
> I'd like to release Gaffer 0.41 and the latest Cortex as a sort of "pre-Arnold 5 checkpoint" before we merge to master and release Gaffer 0.42 with Arnold 5 support. Does that sound OK?

Sure.

> I think I also mentioned that I structured the Encapsulate work in preparation for an API change in Cortex that passes node names and parents to the NodeAlgo methods - could you take a look at that?

I'm a bit confused by your wording - at first I thought you meant you wanted me to take a look at work you've done - but I don't see any PRs.  So maybe what you mean is that I should do that work as part of the Arnold 5 switch in Cortex?  That makes sense.

danieldresser-ie commented:
Fixed the name of the light_path_expressions variable.

As for commit messages, I can do some cleanup before I do PRs on master - I ( perhaps a bit preemptively ) ran through taking a look at the rebase to master on Friday to get a feel for how it interacted with the encapsulate stuff, so if we're this close to getting this merged as is, I might see if I can get away without redoing that rebase.

danieldresser-ie commented:
Err, I meant in terms of general length of commit message subjects I might not worry about cleaning it up until I go to master.  I did go back and clear out the inaccurate old comment.

johnhaddon commented:
> I tried to figure out how the AiEnum approach would work, but it looks like Renderer::ArnoldAtttributes gets repeatedly constructed for each location anyway

I have been wondering about caching AttributesInterfaces in the InteractiveRender node, so we reuse the interface for all locations with the same attributes. For the new 3Delight backend I'm actually doing the caching/sharing internal to the renderer, but maybe it makes sense to bring it out. But anyway, something for later I think.

> The comment on enum plugs comes from talking to Don about it, and him pointing out that aside from performance, a string really isn't a very clean representation of something that can only have 3 values.

What does Don propose instead? Bear in mind that it needs to be passed to the Renderer as a DataPtr, and a tonne of different `EnumData<SomeEnum>` classes would introduce a huge amount of (public) bloat around what is currently a nice, cleanly hidden implementation behind a single interface. I take your point, but I don't see a solution with better trade offs.

> I can scrap the comment if you prefer.

Yes please - could you do that and then go ahead and merge?

> I don't see much problem mentioning plugs in a comment - is there any point in pretending that anyone cares about Cortex renderer backends in any context other than Gaffer?

Hradec expressed reluctance to lose his python procedurals, and my aim is to provide that capability for the new renderer backends too, so yes. But more generally, the Renderer interface is a primarily an abstraction to insulates the backend from Gaffer - even if Gaffer _is_ the main client, it's still useful to think of them separately. If you stop "pretending", you lose the abstraction.

>> I think I also mentioned that I structured the Encapsulate work in preparation for an API change in Cortex that passes node names and parents to the NodeAlgo methods - could you take a look at that?

> I'm a bit confused by your wording - at first I thought you meant you wanted me to take a look at work you've done - but I don't see any PRs. So maybe what you mean is that I should do that work as part of the Arnold 5 switch in Cortex? That makes sense.

Sorry for the lack of clarity. I was alluding to the [node creation functions I added to the the Arnold renderer backend](https://github.com/GafferHQ/gaffer/blob/master/src/GafferArnold/IECoreArnoldPreview/Renderer.cpp#L81). These all take additional `const std::string &name, const AtNode *parentNode` arguments that are equivalent to the new `AiNode()` function arguments added in Arnold 5.

We need to adjust the IECoreArnold::NodeAlgo API so it takes those additional arguments and passes them through to `AiNode()`. Once that is done, the wrappers I added in the Arnold renderer backend can be removed, and we can just call direct to the Cortex APIs again. I'm hoping that you can do the work of adjusting IECoreArnold::NodeAlgo and removing the wrappers at the same time as preparing your work for master.

> Err, I meant in terms of general length of commit message subjects I might not worry about cleaning it up until I go to master. I did go back and clear out the inaccurate old comment.

Cool. The commit formatting comment was mostly intended for future reference, so if you can do it for master that's great.

danieldresser-ie commented:
Cleaned up comment, merging.
